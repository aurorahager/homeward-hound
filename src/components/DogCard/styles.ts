import styled from '@emotion/styled';
import { Box, Card } from '@mui/material';

export const CardContainer = styled(Card)`
  transition: transform 0.2s ease, box-shadow 0.2s ease;
  &:hover {
    transform: translateY(-5px);
    box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  }
    

  @media (max-width: 787px) {
    width: 85%;
  }
  `

export const CardInfoWrapper = styled(Box)`
display: flex;
flex-direction: column;
        justify-content: space-between;
        height: 100%;
`

export const MediaWrapper = styled(Box)`
width: 100%;
height: 75%;
display: flex;
justify-content: center;
align-items: center;
position: relative;

`