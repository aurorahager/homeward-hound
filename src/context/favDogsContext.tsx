'use client'

import { createContext, ReactNode, useContext, useMemo, useState } from 'react'

interface FavDogsContextType {
  faves: string[]
  setFaves: React.Dispatch<React.SetStateAction<string[]>>
}

const FaveDogsContext = createContext<FavDogsContextType | undefined>(undefined)

export function FavDogsProvider({
  children,
}: {
  children: ReactNode
}): JSX.Element {
  const [faves, setFaves] = useState<string[]>([])

  const value: FavDogsContextType = useMemo(
    () => ({ faves, setFaves }),
    [faves, setFaves],
  )

  return (
    <FaveDogsContext.Provider value={value}>
      {children}
    </FaveDogsContext.Provider>
  )
}

export const useFaveDogs = (): FavDogsContextType => {
  const context = useContext(FaveDogsContext)
  if (context === undefined) {
    throw new Error('useFaveDogs must be used within a FavDogsProvider')
  }
  return context
}
