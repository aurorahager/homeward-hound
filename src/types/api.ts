import {
  AxiosError,
  AxiosRequestConfig,
  AxiosRequestHeaders,
  AxiosResponse,
} from 'axios'

export interface LoginParams {
  name: string
  email: string
}
export interface Dog {
  id: string
  img: string
  name: string
  age: number
  zip_code: string
  breed: string
}
export interface Match {
  match: string
}
export interface SearchResponse {
  resultIds: string[]
  total: number
  next?: string
  prev?: string
}

export interface Error {
  message: string
  status: number
}

export type SWRData<T, K extends string> = {
  [key in K]: T
}

export interface SWRResponse<T, K extends string> extends SWRData<T, K> {
  isLoading: boolean
  isError: Error | null
}

export interface RequestConfig extends AxiosRequestConfig {
  headers?: AxiosRequestHeaders
  params?: Record<string, string | number | boolean>
}

export interface ResponseAxiosError<T = Error> extends AxiosError<T> {
  response?: {
    data: T
    status: number
    statusText: string
    headers: Record<string, string>
    config: unknown
    request?: unknown
  }
}

type Response<T> = SearchResponse | T[] | Match

export interface ApiResponse<T> extends AxiosResponse<Response<T>> {}
