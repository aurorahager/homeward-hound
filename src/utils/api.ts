'use client'

import Axios, {
  InternalAxiosRequestConfig,
  AxiosError,
  AxiosResponse,
} from 'axios'

import { API_BASE_URL } from '@/utils/constants'

export const api = Axios.create({
  baseURL: API_BASE_URL,
})

const requestInterceptor = (config: InternalAxiosRequestConfig): InternalAxiosRequestConfig => {
  const { headers } = config
  if (headers) {
    headers.Accept = 'application/json'
  }

  return {
    ...config,
    headers,
    withCredentials: true,
  }
}

const requestErrorInterceptor = (error: AxiosError): Promise<AxiosError> =>
  Promise.reject(error)

const responseInterceptor = (
  response: AxiosResponse
): AxiosResponse => response.data

const responseErrorInterceptor = (
  error: AxiosError,
): Promise<AxiosError> => {
  return Promise.reject(error)
}

api.interceptors.request.use(requestInterceptor, requestErrorInterceptor)
api.interceptors.response.use(responseInterceptor, responseErrorInterceptor)
