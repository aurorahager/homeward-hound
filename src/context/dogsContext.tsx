'use client'

import React, {
  ReactNode,
  createContext,
  useContext,
  useMemo,
  useReducer,
} from 'react'

import { Context, DogAction, DogState } from '@/types/ui'

const initialState: DogState = {
  favoriteIds: [],
  matchId: null,
  isAuthenticated: false,
  prevPage: '',
  nextPage: '',
  query: '',
}

const dogReducer = (state: DogState, action: DogAction): DogState => {
  switch (action.type) {
    case 'SET_FAVORITE':
      return {
        ...state,
        favoriteIds: state.favoriteIds.includes(action.payload)
          ? state.favoriteIds.filter((id) => id !== action.payload)
          : [...state.favoriteIds, action.payload],
      }
    case 'SET_MATCH':
      return { ...state, matchId: action.payload }
    case 'SET_PAGES':
      return {
        ...state,
        prevPage: action.payload.prev,
        nextPage: action.payload.next,
      }
    case 'SET_SEARCH_QUERY':
      return { ...state, query: action.payload }
    case 'LOGIN':
      return { ...state, isAuthenticated: true }
    case 'LOGOUT':
      return { ...state, isAuthenticated: false }
    default:
      return state
  }
}

const DogContext = createContext<
  { state: DogState; dispatch: React.Dispatch<DogAction> } | undefined
>(undefined)

export function DogProvider({ children }: { children: ReactNode }): ReactNode {
  const [state, dispatch] = useReducer(dogReducer, initialState)

  const value = useMemo(() => ({ state, dispatch }), [state, dispatch])
  return <DogContext.Provider value={value}>{children}</DogContext.Provider>
}

export const useDogContext = (): Context => {
  const context = useContext(DogContext)
  if (!context) {
    throw new Error('useDogContext outside of the DogProvider')
  }
  return context
}
