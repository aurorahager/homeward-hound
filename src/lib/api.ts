import Axios, { AxiosError } from 'axios'

import {
  ApiResponse,
  Dog,
  RequestConfig,
  ResponseAxiosError,
  SearchedDogs,
} from '@/types/api'

import { API_BASE_URL } from '@/utils/constants'

export const api = Axios.create({
  baseURL: API_BASE_URL,
})

const requestInterceptor = (config: RequestConfig): RequestConfig => {
  console.log('PLEASEEEE', config)
  const { headers } = config
  if (headers) {
    headers.Accept = 'application/json'
  }

  return {
    ...config,
    headers,
    withCredentials: true,
  }
}

const requestErrorInterceptor = (error: AxiosError): Promise<AxiosError> => {
  console.error(`[request error] [${error.message} - ${error.status}]`)
  return Promise.reject(error)
}

const responseInterceptor = (
  response: ApiResponse<Dog[] | SearchedDogs | string[]>,
): ApiResponse<Dog[] | SearchedDogs | string[]> => {
  console.info(`[response] [${JSON.stringify(response.data)}]`)
  return response.data
}

const responseErrorInterceptor = (
  error: ResponseAxiosError,
): Promise<ResponseAxiosError> => {
  console.error(`[response error] [${error.message} - ${error.status}]`)
  // const message = error.response?.data?.message || error.message
  // TODO: show error notification
  if (error.response?.status === 401) {
    // TODO: redirect to login page
  }

  return Promise.reject(error)
}

api.interceptors.request.use(requestInterceptor, requestErrorInterceptor)
api.interceptors.response.use(responseInterceptor, responseErrorInterceptor)
