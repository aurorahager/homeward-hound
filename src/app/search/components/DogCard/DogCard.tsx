'use client'

import { FavoriteBorderOutlined, FavoriteOutlined } from '@mui/icons-material'
import {
  Box,
  Button,
  Card,
  CardActions,
  CardContent,
  CardMedia,
  Typography,
} from '@mui/material'

import { useFaveDogs } from '@/context/favDogsContext'
import { Dog } from '@/types/api'

export default function DogCard({ dog }: { dog: Dog }) {
  console.log('DOG INFO', dog)
  const { faves, setFaves } = useFaveDogs()

  const handleFavorite = () => {
    setFaves((prev) => [...prev, dog.id])
    console.log('FAVES', faves)
  }
  return (
    <Card sx={{ width: 345 }}>
      <Box
        sx={{
          width: '100%',
          height: 250,
          overflow: 'hidden',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <CardMedia
          alt={`${dog.name} the ${dog.breed}`}
          component="img"
          image={dog.img}
          sx={{
            height: '100%',
            width: '100%',
            objectFit: 'cover',
            filter: 'sepia(20%) contrast(90%) brightness(80%)',
            '&::after': {
              content: '""',
              position: 'absolute',
              top: 0,
              left: 0,
              width: '100%',
              height: '100%',
              background: 'rgba(210, 180, 140, 0.35)',
              mixBlendMode: 'multiply',
              pointerEvents: 'none',
            },
          }}
        />
      </Box>
      <CardContent>
        <Typography gutterBottom component="div" variant="h5">
          {dog.name}
        </Typography>
        <Typography sx={{ color: 'text.secondary' }} variant="body2">
          {dog.breed} {dog.age} {dog.zip_code}
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small" onClick={handleFavorite}>
          {faves.includes(dog.id) ? (
            <FavoriteOutlined />
          ) : (
            <FavoriteBorderOutlined />
          )}
          Favorite
        </Button>
      </CardActions>
    </Card>
  )
}
