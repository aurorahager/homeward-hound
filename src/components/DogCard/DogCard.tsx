'use client'

import { FavoriteOutlined, Brightness1 } from '@mui/icons-material'
import {
  Box,
  IconButton,
  CardContent,
  CardMedia,
  Typography,
} from '@mui/material'

import { CardContainer, CardInfoWrapper, MediaWrapper } from './styles'

import { useDogContext } from '@/context/dogsContext'
import { Dog } from '@/types/api'


export default function DogCard({ dog }: { dog: Dog }): React.ReactElement {
  const { state, dispatch } = useDogContext()
  const { img, name, breed, zip_code, id, age } = dog ?? {}

  const handleFavorite = () => {
    dispatch({ type: 'SET_FAVORITE', payload: dog.id })
  }

  return (
    <CardContainer>
      <CardInfoWrapper>
        <MediaWrapper
        >
          <CardMedia
            alt={`${dog.name} the ${dog.breed}`}
            component="img"
            image={img}
            sx={{
              height: '100%',
              width: '100%',
              objectFit: 'cover',


            }}
          />
          <IconButton aria-label="Favorite" size="small" onClick={handleFavorite} color="secondary" sx={{
            position: 'absolute',
            top: 3,
            right: 20,
          }}>
            {state.favoriteIds.includes(id) ? (
              <FavoriteOutlined sx={{ fontSize: '1.8rem' }} />
            ) : (
              <FavoriteOutlined sx={{
                fontSize: '1.8rem',
                color: 'white'
              }} />
            )}

          </IconButton>
        </MediaWrapper>
        <Box display="flex" sx={{ flexDirection: 'column', justifyContent: 'flex-end', marginBottom: '10%' }}>
          <CardContent component="div" sx={{ display: 'flex', flexDirection: 'column', justifyContent: 'space-between' }}>
            <Box display="flex" alignItems="center">
              <Typography component="h2" variant="h5" sx={{ fontWeight: 800 }}>
                {name}
              </Typography>
              <Brightness1 sx={{ fontSize: '0.5rem', mx: '0.4rem' }} />
              <Typography variant="h6" component="span">
                {breed}
              </Typography>
            </Box>
            <Typography sx={{ color: 'text.secondary' }} variant="subtitle1">
              Age: {age} Location: {zip_code}
            </Typography>
          </CardContent>
        </Box>
      </CardInfoWrapper>
    </CardContainer>
  )
}
