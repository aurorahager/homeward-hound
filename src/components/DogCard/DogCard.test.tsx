import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import { useDogContext } from '@/context/dogsContext'
import { Dog } from '@/types/api'

import DogCard from './DogCard'

jest.mock('@/context/dogsContext')

describe('DogCard', () => {
  const dog: Dog = {
    id: '1',
    img: 'https://example.com/dog.jpg',
    name: 'Buddy',
    breed: 'Golden Retriever',
    zip_code: '12345',
    age: 5,
  }

  const mockDispatch = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
    ;(useDogContext as jest.Mock).mockReturnValue({
      state: { favoriteIds: [] },
      dispatch: mockDispatch,
    })
  })

  it('renders dog image', () => {
    render(<DogCard dog={dog} />)
    const image = screen.getByAltText('Buddy the Golden Retriever')

    expect(image).toBeInTheDocument()
  })

  it('renders placeholder image when img is undefined', () => {
    render(<DogCard dog={{ ...dog, img: undefined as unknown as string }} />)
    const placeholderImage = screen.getByAltText('Buddy the Golden Retriever')

    expect(placeholderImage).toHaveAttribute(
      'src',
      'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==',
    )
  })

  it('renders dog name', () => {
    render(<DogCard dog={dog} />)
    const name = screen.getByText('Buddy')

    expect(name).toBeInTheDocument()
  })

  it('renders dog breed', () => {
    render(<DogCard dog={dog} />)
    const breed = screen.getByText('Golden Retriever')

    expect(breed).toBeInTheDocument()
  })

  it('renders dog age and location', () => {
    render(<DogCard dog={dog} />)
    const details = screen.getByText('Age: 5 Location: 12345')

    expect(details).toBeInTheDocument()
  })

  it('renders favorite icon', () => {
    render(<DogCard dog={dog} />)
    const favoriteIcon = screen.getByLabelText('Favorite')

    expect(favoriteIcon).toBeInTheDocument()
  })

  it('triggers dispatch when favorite icon is clicked', () => {
    render(<DogCard dog={dog} />)
    const favoriteIcon = screen.getByLabelText('Favorite')
    fireEvent.click(favoriteIcon)

    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'SET_FAVORITE',
      payload: dog.id,
    })
  })

  it('shows orange favorite icon when dog is favorited', () => {
    ;(useDogContext as jest.Mock).mockReturnValue({
      state: { favoriteIds: ['1'] },
      dispatch: mockDispatch,
    })
    const { container } = render(<DogCard dog={dog} />)
    expect(container).toMatchSnapshot()
  })

  it('shows white favorite icon when dog is not favorited', () => {
    render(<DogCard dog={dog} />)
    const heartIcon = screen.getByLabelText('Favorite').querySelector('svg')

    expect(heartIcon).toHaveStyle({ color: 'white' })
  })
})
