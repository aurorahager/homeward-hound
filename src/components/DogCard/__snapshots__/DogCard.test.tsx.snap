// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogCard shows orange favorite icon when dog is favorited 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-z89q2r-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiBox-root css-1hcrhda"
    >
      <div
        class="MuiBox-root css-qmznoe"
      >
        <img
          alt="Buddy the Golden Retriever"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-ak1pqf-MuiCardMedia-root"
          src="https://example.com/dog.jpg"
        />
        <button
          aria-label="Favorite"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall css-1dr8kgd-MuiButtonBase-root-MuiIconButton-root"
          id=":r7:"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-3wu62u-MuiSvgIcon-root"
            data-testid="FavoriteOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54z"
            />
          </svg>
        </button>
      </div>
      <div
        class="MuiBox-root css-1vimsrb"
      >
        <div
          class="MuiCardContent-root css-18jh05-MuiCardContent-root"
        >
          <div
            class="MuiBox-root css-fv3lde"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5 css-efm3u7-MuiTypography-root"
            >
              Buddy
            </h2>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-8gpknc-MuiSvgIcon-root"
              data-testid="Brightness1Icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
            </svg>
            <span
              class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
            >
              Golden Retriever
            </span>
          </div>
          <div
            class="MuiTypography-root MuiTypography-subtitle1 css-1t4hig0-MuiTypography-root"
          >
            Age: 
            5
             Location: 
            12345
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
